### PR TITLE
Cartesian: Refactoring repository layout and package organization

### PR DESCRIPTION
**Pull request by egparedes**
`gt4py` -> `gt4py.cartesian`
`eve` -> `gt4py.eve`
`gtc` -> `gt4py.cartesian.gtc`
`gt4py.storage` -> `gt4py.storage`
 
Tests split into sub-packages and
`test_unittests` -> `unit_tests`
`test_integration` -> `integration_tests`
 
Additionally: Refactored `gt4py.storage` does not depend on `gt4py.cartesian`.

### CODE REVIEW

**Comment by jdahm** [comment]
Resolves #1040.

**Comment by jdahm** [comment]
Looks like there are a few references to `gt4py.config` in comments that still need updating 👍

**Comment by egparedes** [comment]
At this point, all CPU tests (including DaCe) are passing in my laptop. There are still several rough edges, though, which should be solved either in this PR or in following ones:

- typing errors need to be solved
- autodocs need to work properly
-  imports in subpackage's  `__init__.py`s need to be polished and improved to follow some coherent policy in all subpackages
- `tox` config should be aligned with the version in the `functional` branch and `tox` itself updated to version 4 (which also involves some changes in the configuration)
- `pre-commit` settings should be aligned with the version in `functional` branch
- package and tools configurations should be moved to `pyproject.toml` (ideally switching to a modern build backend like `hatchling`)

**Comment by havogt** [comment]
bors try

**Comment by havogt** [comment]
bors try

**Comment by havogt** [comment]
bors try

File: src/gt4py/storage/layout.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+CUDALayout: Final[LayoutInfo] = {
+    "alignment": 32,
+    "device": "gpu",
+    "layout_map": make_cuda_layout_map,
+    "is_optimal_layout": layout_checker_factory(make_cuda_layout_map),
+}
+register("cuda", CUDALayout)
+
+GPULayout: Final[LayoutInfo] = CUDALayout
+register("gpu", GPULayout)
```
Comment (gronerl) [inline src/gt4py/storage/layout.py:164]: What's the idea/principle for only defining non-dace backends' layouts here while leaving the dace ones in-line withe backends?

**Review by gronerl — APPROVED** [review]

File: src/gt4py/storage/layout.py (hunk 1 above)
Comment (egparedes) [reply src/gt4py/storage/layout.py:164]: There are not super strong reasons, just that dace is an optional dependency and I find it slightly weird to define layouts for backends that might have not been installed.

**Pull request merged**